Graphics/GraphicsEngineWebGPU/src/BufferWebGPUImpl.cpp

```diff
@@ -350,6 +261,53 @@ void BufferWebGPUImpl::SetDynamicAllocation(DeviceContextIndex CtxId, DynamicMem
     m_DynamicAllocations[CtxId] = std::move(Allocation);
 }
 
+const BufferWebGPUImpl::StagingBufferSyncInfo* BufferWebGPUImpl::GetStagingBufferInfo()
+{
+    VERIFY(m_Desc.Usage == USAGE_STAGING, "Staging buffer is expected");
+
+    if (m_Desc.CPUAccessFlags & CPU_ACCESS_READ)
+        return FindAvailableReadMemoryBuffer();
+    if (m_Desc.CPUAccessFlags & CPU_ACCESS_WRITE)
+        return FindAvailableWriteMemoryBuffer();
+
+    UNEXPECTED("Unexpected CPU access flags");
+    return nullptr;
+}
+
+void BufferWebGPUImpl::FlushPendingWrites(Uint32 BufferIdx)
+{
+    VERIFY(m_Desc.Usage == USAGE_STAGING, "Staging buffer is expected");
+    VERIFY(m_Desc.CPUAccessFlags & CPU_ACCESS_WRITE, "Unexpected CPUAccessFlags");
+
+    const auto& BufferInfo = m_StagingBufferInfo[BufferIdx];
+
+    void* pData = wgpuBufferGetMappedRange(BufferInfo.wgpuBuffer, 0, WGPU_WHOLE_MAP_SIZE);
+    memcpy(pData, m_MappedData.data(), m_MappedData.size());
+    wgpuBufferUnmap(BufferInfo.wgpuBuffer);
+    m_StagingBufferInfo.clear();
+}
+
+void BufferWebGPUImpl::ProcessAsyncReadback(Uint32 BufferIdx)
+{
+    auto MapAsyncCallback = [](WGPUBufferMapAsyncStatus MapStatus, void* pUserData) {
+        if (MapStatus != WGPUBufferMapAsyncStatus_Success && MapStatus != WGPUBufferMapAsyncStatus_DestroyedBeforeCallback)
+            DEV_ERROR("Failed wgpuBufferMapAsync: ", MapStatus);
+
+        if (MapStatus == WGPUBufferMapAsyncStatus_Success && pUserData != nullptr)
+        {
+            auto* pBufferInfo = static_cast<StagingBufferSyncInfo*>(pUserData);
+
+            const auto* pData = static_cast<const uint8_t*>(wgpuBufferGetConstMappedRange(pBufferInfo->wgpuBuffer, 0, WGPU_WHOLE_MAP_SIZE));
```

Comment:
⚠️ This is the source of potential problem: the data passed to the callback must be alive and potentially outlive the buffer itself.
To ensure this:
- Make the data derived from IObject
- Manually call AddRef before passing data to the callback
- Manually call Release in the callback

Reply:
There's no problem with the lifetime here, because when the buffer is destroyed, a WGPUBufferMapAsyncStatus_DestroyedBeforeCallback event will be triggered.

Reply:
UPD:
I discovered a situation where `WGPUBufferMapAsyncStatus_DestroyedBeforeCallback` won't help us. When we use a `WGPUBuffer` in a `WGPUCommandList`, it implicitly increases the reference count of the object, which could lead to a situation where we receive `WGPUBufferMapAsyncStatus_Success`, but the object has already been deleted on our side. The method you suggested won't work because we also need to track the lifetime of the object itself, as `pMappedData`might be invalid. Therefore, I decided to simply store a reference to the object. This way, we can just increase the reference count when the asynchronous operation is called and decrease it when the operation is completed.

Reply:
Here is what I suggest:
Before the callback:
```
pData->AddRef();
Callback() = [](){
  // Do work
  pData->Release();
};
Callback(... pData)
```
`pData` **must** contain all data that may be accessed by the callback. This is guaranteed to work,